triton-vm/src/table/ram_table.rs

```diff
@@ -218,79 +218,191 @@ impl RamTable {
     ) {
+        profiler!(start "ram table");
         assert_eq!(BASE_WIDTH, base_table.ncols());
         assert_eq!(EXT_WIDTH, ext_table.ncols());
         assert_eq!(base_table.nrows(), ext_table.nrows());
 
-        let mut running_product_for_perm_arg = PermArg::default_initial();
-        let mut clock_jump_diff_lookup_log_derivative = LookupArg::default_initial();
+        let extension_column_indices = RamExtTableColumn::iter()
+            // RunningProductOfRAMP + FormalDerivative are constitute one
+            // slice and are populated by the same function
+            .filter(|column| *column != RamExtTableColumn::FormalDerivative)
+            .map(|column| column.ext_table_index())
+            .collect_vec();
+        let extension_column_slices = horizontal_multi_slice_mut(
+            ext_table.view_mut(),
+            &contiguous_column_slices(&extension_column_indices),
+        );
+        let extension_functions = [
+            Self::extension_column_running_product_of_ramp_and_formal_derivative,
+            Self::extension_column_bezout_coefficient_0,
+            Self::extension_column_bezout_coefficient_1,
+            Self::extension_column_running_product_perm_arg,
+            Self::extension_column_clock_jump_difference_lookup_log_derivative,
+        ];
+        extension_functions
+            .into_par_iter()
+            .zip_eq(extension_column_slices)
+            .for_each(|(generator, slice)| {
+                generator(base_table, challenges).move_into(slice);
+            });
+
+        profiler!(stop "ram table");
+    }
 
-        // initialize columns establishing Bézout relation
+    fn extension_column_running_product_of_ramp_and_formal_derivative(
+        base_table: ArrayView2<BFieldElement>,
+        challenges: &Challenges,
+    ) -> Array2<XFieldElement> {
         let bezout_indeterminate = challenges[RamTableBezoutRelationIndeterminate];
-        let clock_jump_difference_lookup_indeterminate =
-            challenges[ClockJumpDifferenceLookupIndeterminate];
+        let row_count = base_table.nrows();
+
+        let mut extension_columns = Vec::with_capacity(2 * row_count);
         let mut running_product_ram_pointer =
             bezout_indeterminate - base_table.row(0)[RamPointer.base_table_index()];
         let mut formal_derivative = xfe!(1);
-        let mut bezout_coefficient_0 =
-            base_table.row(0)[BezoutCoefficientPolynomialCoefficient0.base_table_index()].lift();
-        let mut bezout_coefficient_1 =
-            base_table.row(0)[BezoutCoefficientPolynomialCoefficient1.base_table_index()].lift();
-
-        let mut previous_row: Option<ArrayView1<BFieldElement>> = None;
-        for row_idx in 0..base_table.nrows() {
-            let current_row = base_table.row(row_idx);
-            let clk = current_row[CLK.base_table_index()];
+        let mut previous_row: Option<ArrayBase<ViewRepr<&BFieldElement>, Dim<[usize; 1]>>> = None;
```

Comment:
Please use the alias `ArrayView1<_>`.

```rust
let mut previous_row = None::< /* here */>;
```

Reply:
Arguably even cleaner is to not drag the mutable `Option<ArrayView1<_>>` around in the first place, since the only case in which something happens is if that `Option<_>` is `Some(_)`. An approach using `base_table.rows().into_iter().tuple_windows()` comes to mind:

```rust
let num_rows = base_table.nrows();
let indeterminate = challenges[RamTableBezoutRelationIndeterminate];

let mut extension_columns = Vec::with_capacity(2 * num_rows);
let mut running_product_ram_pointer =
    indeterminate - base_table.row(0)[RamPointer.base_table_index()];
let mut formal_derivative = xfe!(1);

extension_columns.push(running_product_ram_pointer);
extension_columns.push(formal_derivative);
for (previous_row, current_row) in base_table.rows().into_iter().tuple_windows() {
    let is_no_padding_row =
        previous_row[InstructionType.base_table_index()] != PADDING_INDICATOR;
    let current_ram_pointer = current_row[RamPointer.base_table_index()];
    let previous_ram_pointer = previous_row[RamPointer.base_table_index()];
    if is_no_padding_row && previous_ram_pointer != current_ram_pointer {
        formal_derivative = (indeterminate - current_ram_pointer) * formal_derivative
            + running_product_ram_pointer;
        running_product_ram_pointer *= indeterminate - current_ram_pointer;
    }

    extension_columns.push(running_product_ram_pointer);
    extension_columns.push(formal_derivative);
}
Array2::from_shape_vec((num_rows, 2), extension_columns).unwrap()
```

Reply:
Adopted in eb3ba865.